modules/packages/Crypto.chpl

```diff
@@ -0,0 +1,953 @@
+       in the array. This kind of use case, is useful specifically in developing
+       one-to-many systems such as GPG.
+
+       For instance, consider a server-client model where the server performs
+       encryption using the public keys of all the clients and returns an `Envelope`
+       to each client. Now, the dencryption can only be performed by each client
+       using only their own private key.
+
+       :arg plaintext: A `CryptoBuffer` representing the plaintext to be
+                       encrypted.
+       :type plaintext: `CryptoBuffer`
+
+       :arg keys: An array of `RSAKey` objects. This depends on the number of
+                  different RSA encryptions that are to be performed on the same
+                  plaintext.
+       :type keys: `[] RSAKey`
+
+       :return: An `Envelope` object which comprises of the IV buffer, array of
+                RSA encrypted keys and AES encrypted ciphertext.
+       :rtype: `Envelope`
+
+    */
+    proc encrypt(plaintext: CryptoBuffer, keys: [] RSAKey): Envelope {
+      var ivLen = EVP_CIPHER_iv_length(EVP_aes_256_cbc()) - 1;
+      var iv: [0..(ivLen: int(64))] uint(8);
+
+      var encSymmKeys: [keys.domain] CryptoBuffer;
+
+      var ciphertext = rsaEncrypt(keys, plaintext, iv, encSymmKeys);
+
+      var envp = new Envelope(new CryptoBuffer(iv), encSymmKeys, new CryptoBuffer(ciphertext));
+      return envp;
+    }
+
+    /* This is the 'RSA' decrypt routine that decrypts the ciphertext buffer. This
+       uses `AES` dencryption to decrypt the ciphertext.
+
+       It's accepts the entire `Envelope` object but only a single `RSAKey` object
+       for unilateral dencryption.
+
+       :arg envp: A `CryptoBuffer` representing the plaintext to be
+                       encrypted.
+       :type envp: `Envelope`
+
+       :arg key: An array of `RSAKey` objects. This depends on the number of
+                  different RSA encryptions that are to be performed on the same
+                  plaintext.
+       :type key: `RSAKey`
+
+       :return: An `Envelope` object which comprises of the IV buffer, array of
+                RSA encrypted keys and AES encrypted ciphertext.
+       :rtype: `CryptoBuffer`
+
+    */
+    proc decrypt(envp: Envelope, key: RSAKey): CryptoBuffer {
+      var iv = envp.getIV().getBuffData();
+      var ciphertext = envp.getEncMessage().getBuffData();
+      var encKeys = envp.getEncKeys();
+
+      var plaintext = rsaDecrypt(key, iv, ciphertext, encKeys);
+      var plaintextBuff = new CryptoBuffer(plaintext);
+      return plaintextBuff;
+    }
+  }
+}
+
+
+/*
+    Support for low-level native C_OpenSSL bindings.
+    This submodule wraps the C_OpenSSL implementation, providing access to
+    most of the C_OpenSSL calls.
+
+    Refer to the
+    `C_OpenSSL documentation <https://www.openssl.org/docs/manmaster/man3/>`_
+    of the reference version for the usage of this module.
+    
+  */
+module C_OpenSSL {
```

Comment:
This is actually a separate top-level module (which shows up in the top-level TOC when building docs). To make it a submodule, put it within the scope of `module Crypto { ... }`.

Reply:
@saru95 - this got buried in a recent commit, but I want to make sure you see it.